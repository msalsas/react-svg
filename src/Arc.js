var React = require('react');
var PropTypes = React.PropTypes;
var Component = React.Component;

class Arc extends Component {
    static propTypes = {
        radius: PropTypes.number,
        degreesOffset: PropTypes.number,
        degrees: PropTypes.number,
        strokeWidth: PropTypes.number,
        strokeColor: PropTypes.string,
        fillColor: PropTypes.string
    };
	static defaultProps = {
		radius: 50,
		degreesOffset: 45,
		degrees: 360,
		strokeWidth: 5,
		strokeColor: '#fff',
        fillColor: 'none'
    };
    render() {
        return (
            <svg width={this.props.radius * 2 + this.props.strokeWidth * 2} height={this.props.radius * 2 + this.props.strokeWidth * 2} xmlns="http://www.w3.org/2000/svg">
                <g>
                    <path d={this.describeArc(this.props.radius, this.props.strokeWidth, this.props.degreesOffset, this.props.degrees)} className="msalsas-arc"
                    fill={this.props.fillColor} strokeWidth={this.props.strokeWidth} stroke={this.props.strokeColor}/>
                </g>
            </svg>
        );
    }

    polarToCartesian = function(radius, angleInDegrees) {
        var angleInRadians = (angleInDegrees - 90) * Math.PI / 180.0;

        return {
            x: radius + (radius * Math.cos(angleInRadians)),
            y: radius + (radius * Math.sin(angleInRadians))
        };
    };

    describeArc = function(radius, strokeWidth, startAngle, endAngle) {
        endAngle = endAngle === 360 ? 359.9 : endAngle;
        var start = this.polarToCartesian(radius, endAngle);
        var end = this.polarToCartesian(radius, startAngle);

        var largeArcFlag = endAngle - startAngle <= 180 ? "0" : "1";

        return [
            "M", start.x + strokeWidth, start.y + strokeWidth,
            "A", radius, radius, 0, largeArcFlag, 0, end.x + strokeWidth, end.y + strokeWidth
        ].join(" ");
    };
}

export default Arc;
