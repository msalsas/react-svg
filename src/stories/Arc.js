import React from 'react';
import { storiesOf, action } from '@storybook/react';
import Arc from '../Arc';

storiesOf('Arc', module)
    .add('orange 270ยบ, 80px radius', () => {
        const props = {
            radius: 80,
            degreesOffset: 0,
            degrees: 270,
            strokeWidth: 3,
            strokeColor: 'orange',
            fillColor: 'none'
        };

        return renderArc(props);
    })
    .add('yellow 70ยบ, 100px radius, 50ยบ offset', () => {
        const props = {
            radius: 100,
            degreesOffset: 50,
            degrees: 70,
            strokeWidth: 3,
            strokeColor: 'yellow',
            fillColor: 'none'
        };

        return renderArc(props);
    })
    .add('red circle', () => {
        const props = {
            radius: 50,
            degreesOffset: 0,
            degrees: 360,
            strokeWidth: 1,
            strokeColor: 'red',
            fillColor: 'red'
        };

        return renderArc(props);
    })
    .add('without options', () => {
        const props = {};

        return renderArc(props);
    });


function renderArc(props) {
    return (
        <div className="svg-wrapper" style={{ background: 'gray' }}>
            <Arc {...props} />
        </div>
    );
}