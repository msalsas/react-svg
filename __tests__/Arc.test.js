import React from 'react';
import Arc from '../src/Arc';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

describe('Test Arc component', () => {
    const props = {
        radius       : 80,
        degreesOffset: 0,
        degrees      : 270,
        strokeWidth  : 20,
        strokeColor  : 'orange',
        fillColor    : 'none',
        strokeLinecap: 'butt'
    };
    const arc = shallow((<Arc {...props}/>));

    it('should be defined', () => {
        expect(arc).toBeDefined();
    });
    it('should render correctly', () => {
        expect(arc).toMatchSnapshot();
    });
});
