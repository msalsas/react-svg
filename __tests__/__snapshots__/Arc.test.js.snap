// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Arc component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Arc
    degrees={270}
    degreesOffset={0}
    fillColor="none"
    radius={80}
    strokeColor="orange"
    strokeLinecap="butt"
    strokeWidth={20}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <g>
        <path
          className="msalsas-arc"
          d="M 20 100.00000000000001 A 80 80 0 1 0 100 20"
          fill="none"
          stroke="orange"
          strokeLinecap="butt"
          strokeWidth={20}
        />
      </g>,
      "height": 200,
      "width": 200,
      "xmlns": "http://www.w3.org/2000/svg",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <path
          className="msalsas-arc"
          d="M 20 100.00000000000001 A 80 80 0 1 0 100 20"
          fill="none"
          stroke="orange"
          strokeLinecap="butt"
          strokeWidth={20}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "msalsas-arc",
          "d": "M 20 100.00000000000001 A 80 80 0 1 0 100 20",
          "fill": "none",
          "stroke": "orange",
          "strokeLinecap": "butt",
          "strokeWidth": 20,
        },
        "ref": null,
        "rendered": null,
        "type": "path",
      },
      "type": "g",
    },
    "type": "svg",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <g>
          <path
            className="msalsas-arc"
            d="M 20 100.00000000000001 A 80 80 0 1 0 100 20"
            fill="none"
            stroke="orange"
            strokeLinecap="butt"
            strokeWidth={20}
          />
        </g>,
        "height": 200,
        "width": 200,
        "xmlns": "http://www.w3.org/2000/svg",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <path
            className="msalsas-arc"
            d="M 20 100.00000000000001 A 80 80 0 1 0 100 20"
            fill="none"
            stroke="orange"
            strokeLinecap="butt"
            strokeWidth={20}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "msalsas-arc",
            "d": "M 20 100.00000000000001 A 80 80 0 1 0 100 20",
            "fill": "none",
            "stroke": "orange",
            "strokeLinecap": "butt",
            "strokeWidth": 20,
          },
          "ref": null,
          "rendered": null,
          "type": "path",
        },
        "type": "g",
      },
      "type": "svg",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
